### ISSUE TITLE
IQueue Split heal OOME

### ISSUE THREAD

**Issue by Danny-Hazelcast** [hazelcast/hazelcast#11937]
failed migration coring during the test, https://github.com/hazelcast/hzCmd-bench/tree/master/lab/hz/split/queue

running on https://github.com/hazelcast/hazelcast/pull/11749

however,  taking the test further reveals a OOME on the small joining cluster side,  
both members in cluster AA failed with  hprof's
the GC pattern looks like a leak, and occurred after a few split join iteration
http://54.82.84.143/~jenkins/workspace/split-x1/3.10-SNAPSHOT/2017_12_07-10_32_26/queue/gc.html

```
hprof is full of `QueueItem` instances and their `HeapData` payload
if this also happens with an `IMap` in the current master -> general split brain issue
if not, it could be a bug in my merger
```

<img width="862" alt="screen shot 2017-12-07 at 15 20 34" src="https://user-images.githubusercontent.com/5988678/33752190-48f9c7fc-dc06-11e7-9ee2-901c372fe5b0.png">

Danny Conlon [3:58 PM] 
http://54.82.84.143/~jenkins/workspace/split-x1/3.10-SNAPSHOT/2017_12_07-10_32_26/queue/gc.html shows leak pattern, and cluster BB is good

A run with larger heap https://hazelcast-l337.ci.cloudbees.com/view/split/job/split-x1/15/console Queue split heal 
show leak pattern http://54.82.84.143/~jenkins/workspace/split-x1/3.10-SNAPSHOT/2017_12_07-13_37_55/queue/gc.html
and failing queue size exceptions.
and same failed migration.
```
Dec 07, 2017 10:46:20 AM com.hazelcast.internal.cluster.impl.ClusterMergeTask
SEVERE: [10.0.0.225]:5701 [HZ] [3.10-SNAPSHOT] While merging...
java.util.concurrent.TimeoutException
```

http://54.82.84.143/~jenkins/workspace/split-x1/3.10-SNAPSHOT/2017_12_07-09_28_46/queue/output/HZ/HzMember5HZBB/out.txt
```
SEVERE: [10.0.0.22]:5701 [HZ] [3.10-SNAPSHOT] While merging...
java.util.concurrent.TimeoutException
    at com.hazelcast.spi.impl.AbstractCompletableFuture.get(AbstractCompletableFuture.java:225)
    at com.hazelcast.internal.cluster.impl.ClusterMergeTask.waitOnFutureInterruptible(ClusterMergeTask.java:169)
    at com.hazelcast.internal.cluster.impl.ClusterMergeTask.executeMergeTasks(ClusterMergeTask.java:152)
    at com.hazelcast.internal.cluster.impl.ClusterMergeTask.run(ClusterMergeTask.java:78)
    at com.hazelcast.instance.LifecycleServiceImpl.runUnderLifecycleLock(LifecycleServiceImpl.java:107)
    at com.hazelcast.internal.cluster.impl.ClusterServiceImpl.merge(ClusterServiceImpl.java:301)
    at com.hazelcast.internal.cluster.impl.operations.MergeClustersOp$1.run(MergeClustersOp.java:63)
    at com.hazelcast.util.executor.CachedExecutorServiceDelegate$Worker.run(CachedExecutorServiceDelegate.java:227)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1152)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:622)
    at java.lang.Thread.run(Thread.java:748)
    at com.hazelcast.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:64)
    at com.hazelcast.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:80)
```

http://54.82.84.143/~jenkins/workspace/split-x1/3.10-SNAPSHOT/2017_12_07-09_28_46/queue/output/HZ/HzMember4HZBB/out.txt
```
INFO: [10.0.0.213]:5701 [HZ] [3.10-SNAPSHOT] Remaining migration tasks in queue => 542
Dec 07, 2017 6:49:32 AM com.hazelcast.internal.partition.InternalPartitionService
INFO: [10.0.0.213]:5701 [HZ] [3.10-SNAPSHOT] Remaining migration tasks in queue => 542
Dec 07, 2017 6:49:44 AM com.hazelcast.internal.partition.impl.MigrationManager
WARNING: [10.0.0.213]:5701 [HZ] [3.10-SNAPSHOT] Migration failed: MigrationInfo{uuid=4ab12aa1-a148-48c9-b979-668393f06010, partitionId=1, source=[10.0.0.144]:5701, sourceUuid=d7bc29cd-25fe-4a80-8ea9-69384ce9c4aa, sourceCurrentReplicaIndex=0, sourceNewReplicaIndex=3, destination=[10.0.0.72]:5701, destinationUuid=bd7041de-f804-4881-a5b1-19710c4cf924, destinationCurrentReplicaIndex=-1, destinationNewReplicaIndex=0, master=[10.0.0.213]:5701, processing=false, status=ACTIVE}
Dec 07, 2017 6:49:47 AM com.hazelcast.internal.partition.impl.MigrationManager
INFO: [10.0.0.213]:5701 [HZ] [3.10-SNAPSHOT] Re-partitioning cluster data... Migration queue size: 543
Dec 07, 2017 6:49:50 AM com.hazelcast.internal.partition.impl.MigrationThread
INFO: [10.0.0.213]:5701 [HZ] [3.10-SNAPSHOT] All migration tasks have been completed, queues are empty.
```

http://54.82.84.143/~jenkins/workspace/split-x1/3.10-SNAPSHOT/2017_12_07-09_28_46/queue/output/HZ/HzMember1HZAA/out.txt is full of
```
WARNING: [10.0.0.198]:5701 [HZ] [3.10-SNAPSHOT] Error while running merge operation: QueueMergeOperation invocation failed to complete due to operation-heartbeat-timeout
```

**Comment by Danny-Hazelcast** [comment]
map split heap on 3.9.1  
https://hazelcast-l337.ci.cloudbees.com/view/split/job/split-x2/2/consoleFull

http://54.82.84.143/~jenkins/workspace/split-x2/3.10-SNAPSHOT/2017_12_07-15_03_59/map/gc.html
also show's leak pattern but on "opposite" larger B cluster 
`can investigate more and create separate issue`

and no map size assertion exceptions

**Comment by mmedenjak** [comment]
On [this run](http://54.82.84.143/~jenkins/workspace/split-x1/3.10-SNAPSHOT/2017_12_07-10_32_26/queue/), two members in smaller cluster fail with OOME: [Member 1](http://54.82.84.143/~jenkins/workspace/split-x1/3.10-SNAPSHOT/2017_12_07-10_32_26/queue/output/HZ/HzMember1HZAA/HzMember1HZAA.hprof), [Member 2](http://54.82.84.143/~jenkins/workspace/split-x1/3.10-SNAPSHOT/2017_12_07-10_32_26/queue/output/HZ/HzMember2HZAA/HzMember2HZAA.hprof). Test code:
```
#!/usr/bin/env bash

hzVersion=${1:-$(hazelcastVersion)}
size=${2:-"M3C3"}
user=${3:-ec2-user}

hzProps="-Dhazelcast.enterprise.license.key=${HAZELCAST_EE_KEY}"
hz memberOps "-Xms100M -Xmx100M ${hzProps}"
hz clientOps "-Xms50M -Xmx50M ${hzProps}"
hz cluster -tag AA -ee -size M2 -v ${hzVersion} -boxes a.box -user ${user}
hz cluster -tag BB -ee -size M3 -v ${hzVersion} -boxes a.box -user ${user}

hz run loadA validateA

for i in {0..30} ; do
    hz split AA BB
    hz run untilClusterSize2  untilClusterSize3

    hz run loadB loadC

    hz heal
    hz run until_AA_merged untilClusterSafe

    hz run validateA
    hz run validateB
    hz run validateC

    hz run clearB clearC
done
```
In [this run](http://54.82.84.143/~jenkins/workspace/split-x1/3.10-SNAPSHOT/2017_12_07-13_37_55/queue/) there was no OOME but the members show a leak pattern. The test code is:
```
#!/usr/bin/env bash

hzVersion=${1:-$(hazelcastVersion)}
size=${2:-"M3C3"}
user=${3:-ec2-user}

hzProps="-Dhazelcast.enterprise.license.key=${HAZELCAST_EE_KEY}"
hz memberOps "-Xms1G -Xmx1G ${hzProps}"
hz clientOps "-Xms50M -Xmx50M ${hzProps}"
hz cluster -tag AA -ee -size M2 -v ${hzVersion} -boxes a.box -user ${user}
hz cluster -tag BB -ee -size M3 -v ${hzVersion} -boxes a.box -user ${user}

hz run loadA validateA

for i in {0..60} ; do
    hz split AA BB
    hz run untilClusterSize2  untilClusterSize3

    hz run loadB loadC

    hz heal
    hz run until_AA_merged untilClusterSafe

    hz run validateA
    hz run validateB
    hz run validateC

    hz run clearB clearC
done
```

**Comment by mmedenjak** [comment]
The issue could be caused by a different number of permits released and acquired by a semaphore in the `Merger` class. [Here](https://github.com/mmedenjak/hazelcast/blob/9714c360a70f5e8aba4a041b813fad899f2e38d6/hazelcast/src/main/java/com/hazelcast/collection/impl/queue/QueueService.java#L468) we acquire `itemCount` permits. But the there will be less permits released - only one per batch. So the code wants to acquire as many permits as there are items but it will release much less - one per batch. 
This should probably just be changed to acquire as many permits as there are batches

So the heap dump has 10 concurrently running instances of `QueueService.Merger`, all blocked on `semaphore.tryAcquire`

**Comment by Donnerbart** [comment]
I hopefully fixed this by adding a separate `operationCount` variable. So we wait for `operationCount` permits, but `itemCount * TIMEOUT_FACTOR` milliseconds. I fixed this in all branches, which seemed to have the same issue.

Thanks a lot for the analysis, Matko!

**Comment by Danny-Hazelcast** [comment]
OOME looks fixed,  however still have failing queue size exceptions.
we can close this issue and created another if preferred

**Comment by Donnerbart** [comment]
Yes, another issue would be great, so it's easier to focus on the remaining problem. It's also easier in the overview, since the OOME title is misleading.
